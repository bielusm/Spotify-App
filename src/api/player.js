import fetchToJson from "./util";

/**
 * [sets currentTrack as the current playing song]
 *
 * @sync
 * @param {string} bearerToken returned from the inital Spotify API call
 * @return {Promise} A track object, keeping the properties
 * {name, artists, external_uris, id, uri, album{name}}
 * for more info on the track object see
 * https://developer.spotify.com/documentation/web-api/reference/object-model/
 */
export const currentPlayer = async bearerToken => {
  try {
    const json = await fetchToJson(
      "https://api.spotify.com/v1/me/player/currently-playing",
      {
        headers: {
          Authorization: "Bearer " + bearerToken
        },
        method: "GET"
      }
    );
    const item = json.item;
    const currentTrack = {
      name: item.name,
      artists: item.artists,
      external_urls: item.external_urls,
      id: item.id,
      uri: item.uri,
      album: {
        name: item.album.name
      }
    };
    return currentTrack;
  } catch (error) {
    if (error.message == 204) throw new Error("No track playing");
    throw error;
  }
};
