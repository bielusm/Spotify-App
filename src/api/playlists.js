import fetchToJson from "./util";

/**
 * Playlist
 * @description A container
 */
export default class Playlist {
  playlists = [];

  /**
   * [Gets information about user playlists and returns them  as a JSON object]
   *
   * @sync
   * @return {[promise]} [Contains playlist object JSON on success]
   *                     and API response on failure
   */
  async getPlaylists(bearerToken) {
    const json = await fetchToJson("https://api.spotify.com/v1/me/playlists", {
      headers: {
        Authorization: "Bearer " + bearerToken
      },
      method: "GET"
    });
    return json;
  }
  /**
   * Takes a ID and name of a playlist and saves/returns all of its tracks as JSON object
   *
   * @async
   * @param {int} ID   [the spotify ID of the playlist used to query the playlist]
   * @param {string} name [used to label the playlist]
   *
   * @return {Promise} Resolves with JSON on success, rejects with Error Object on failure
   */
  async addPlaylistByID(bearerToken, ID, name) {
    if (
      this.playlists.length > 0 &&
      this.playlists.find(element => element.id === ID)
    ) {
      return "already added";
    } else {
      let tracks = [];
      let param =
        "?fields=next,items(track(name,artists,external_urls,id,uri,album(name)))";
      let json = await fetchToJson(
        `https://api.spotify.com/v1/playlists/${ID}/tracks${param}`,
        {
          headers: {
            Authorization: "Bearer " + bearerToken
          },
          method: "GET"
        }
      );
      tracks.push(...json.items);
      const next = json["next"];
      if (next != null) {
        const response = await this.getNextTracks(
          bearerToken,
          ID,
          next,
          tracks
        );
        this.playlists.push({
          id: ID,
          tracks: tracks,
          name: name
        });
      } else {
        this.playlists.push({
          id: ID,
          tracks: tracks,
          name: name
        });
        return "success";
      }
    }
  }

  /**
   * Takes a URL from getNextTracks(ID, tracks) and fetches it,
   * this is used to deal with the fact that Spotify only returns 100 tracks at a time
   *
   * @async
   * @param {number} ID   ID of the Spotify playlist
   * @param {type} next   URL of the next playlist chunk to fetch
   * @param {type} tracks Array containing all tracks so far
   *
   * @return {type} Description
   */
  async getNextTracks(bearerToken, ID, next, tracks) {
    const json = await fetchToJson(next, {
      headers: {
        Authorization: "Bearer " + bearerToken
      },
      method: "GET"
    });
    tracks.push(...json.items);
    next = json["next"];
    if (next != null) return this.getNextTracks(bearerToken, ID, next, tracks);
    else return;
  }

  /**
   * Calls the currentTrackInPlaylist(index, track)
   * function with an index to each playlist, see that function below for more info
   * @param {track object} The track to be checked
   *
   *
   * @return {Array} An array of playlist IDs that contain the currents song
   */
  trackInPlaylists(track) {
    let ids = [];
    let index = 0;
    this.playlists.forEach(playlist => {
      if (this.trackInPlaylist(index, track)) ids.push(playlist);
      index++;
    });
    return ids;
  }

  /**
   * Checks if a playlist at the given index contains the current playing song
   *
   * @param {Number} index The index of the Spotify playlist
   * @param {Track Object} the track to be checked, needs a URI property
   * @return {Boolean} True if the track does contain the playlist, false otherwise
   */
  trackInPlaylist(index, track) {
    let equal = false;
    //TODO forEach is slow for this because it doesen't break the loop
    this.playlists[index].tracks.forEach(item => {
      if (item.track.uri === track.uri) equal = true;
    });
    return equal;
  }

  /**
   *  Adds a given track to a given playlist
   * @async
   * @param {String} playlist_id Spotify URI identifier for the playlist
   * @param {String} track_uri Spotify URI for the track to be added
   *
   * @return {Promise} Success of success and the error on failure
   */
  async addTrackToPlaylist(bearerToken, playlist_id, track) {
    const url = `https://api.spotify.com/v1/playlists/${playlist_id}/tracks?uris=${track.uri}`;
    const op = {
      headers: {
        Authorization: "Bearer " + bearerToken
      },
      method: "POST"
    };
    this.addTrackToCache(playlist_id, track);
    return fetchToJson(url, op);
  }

  /**
   * Takes a playlist URI and removes all instances of a given track URI from it
   * @async
   * @param {String} playlist_id
   * @param {String} track_uri
   *
   * @return {Promise} An empty promise
   */
  async removeTrackFromPlaylist(bearerToken, playlist_id, track) {
    const url = `https://api.spotify.com/v1/playlists/${playlist_id}/tracks`;
    const op = {
      headers: {
        Authorization: "Bearer " + bearerToken,
        "Content-Type": "application/json"
      },
      method: "DELETE",
      body: JSON.stringify({
        tracks: [
          {
            uri: track.uri
          }
        ]
      })
    };
    this.removeTrackFromCache(playlist_id, track);
    return await fetchToJson(url, op);
  }

  /**
   * Removes the playlist with the given playlist URI from the playlist array
   *
   * @param {Number} id  The Spotify URI for the playlist
   */
  removePlaylist(id) {
    const index = this.indexOfPlaylist(id);
    this.playlists.splice(index, 1);
  }

  /**
   * Finds the index of a playlist with the given playlist ID
   *
   * @param {Number} id
   */
  indexOfPlaylist(id) {
    for (let index = 0; index < this.playlists.length; index++) {
      if (this.playlists[index].id === id) return index;
    }
  }

  /**
   * Takes a track object and adds it to the array of all playlists
   *
   * @param {string} playlist_id the Spotify URI of the playlist
   * @param {Spotify track object} track
   */
  addTrackToCache(playlist_id, track) {
    const index = this.indexOfPlaylist(playlist_id);
    this.playlists[index].tracks.push({
      track
    });
  }

  /**
   * Takes a track object and removes it from the array of all playlists
   *
   * @param {string} playlist_id
   * @param {Spotify track object} track
   */
  removeTrackFromCache(playlist_id, track) {
    const index = this.indexOfPlaylist(playlist_id);
    this.playlists[index].tracks = this.playlists[index].tracks.filter(item => {
      return item.track.uri !== track.uri;
    });
  }
}
