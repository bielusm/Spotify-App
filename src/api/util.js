/**
 * @function fetchToJson
 * @description Takes the parameters for a fetch request and converts the response to json
 * @param {string} url URL for the fetch request
 * @param {JSON} op  Options for the fetch request
 *
 * @return {Promise} Resolves with JSON when response.status === 200
 *                   Rejects response object otherwise
 */
export default async (url, op) => {
  const response = await fetch(url, op);
  const status = response.status;
  if (status != 200 && status != 201) {
    throw new Error(response.status);
  }
  const json = await response.json();
  return json;
};
