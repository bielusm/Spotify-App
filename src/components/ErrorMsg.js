import React from "react";
import { connect } from "react-redux";

export const ErrorMsg = props => (
  <>
    {props.errMsg && (
      <h2 id="errMsg" className="my-4">
        {props.errMsg}
      </h2>
    )}
  </>
);

const mapStateToProps = state => ({ errMsg: state.error });

export default connect(mapStateToProps)(ErrorMsg);
