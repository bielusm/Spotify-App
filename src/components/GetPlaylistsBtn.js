import React from "react";

const GetPlaylistsBtn = props => (
  <button
    id="getPlaylistsBtn"
    className="button"
    onClick={props.getPlaylists}
    disabled={props.getPlaylistsDisabled}
  >
    Pick playlists
  </button>
);

export default GetPlaylistsBtn;
