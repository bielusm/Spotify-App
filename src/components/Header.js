import React from "react";

const Header = () => <h1>Spotify Manager</h1>;
export default Header;
