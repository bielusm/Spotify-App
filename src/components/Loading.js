import React from "react";

export const Loading = props => (
  <div>
    {props.loading && (
      <div className="loadingStatus">
        <div className="spinner-border" role="status">
          <span className="sr-only"></span>
        </div>
        <h2 className="mx-2 d-inline">Loading...</h2>
      </div>
    )}
  </div>
);

export default Loading;
