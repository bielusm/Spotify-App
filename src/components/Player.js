import React from "react";
import { Button, IconButton } from "@material-ui/core";

import {
  modifyLibrary,
  startPlayer,
  pausePlayer,
  nextTrack,
  prevTrack
} from "../api/player";

import { setErrorMsg } from "../store/actions/error";
import { connect } from "react-redux";

export class Player extends React.Component {
  togglePause = async () => {
    if (this.props.is_playing)
      pausePlayer(this.props.bearerToken).then(() => this.props.update());
    else startPlayer(this.props.bearerToken).then(() => this.props.update());
  };

  nextTrack = async () => {
    nextTrack(this.props.bearerToken).then(() => this.props.update());
  };

  prevTrack = async () => {
    prevTrack(this.props.bearerToken).then(() => {
      this.props.update();
    });
  };

  toggleLike = async () => {
    modifyLibrary(
      this.props.bearerToken,
      this.props.id,
      this.props.likes_song
    ).then(() => this.props.update());
  };
  render() {
    return (
      <>
        <IconButton
          className="prevBtn fas fa-backward"
          onClick={this.prevTrack}
        />

        <IconButton
          className={
            "playBtn fas " + (this.props.is_playing ? "fa-pause" : "fa-play")
          }
          onClick={this.togglePause}
        />
        <IconButton
          className="nextBtn fas fa-forward"
          onClick={this.nextTrack}
        />
        <IconButton
          className={
            "likeBtn " +
            (this.props.likes_song ? "fas fa-heart" : "far fa-heart")
          }
          onClick={this.toggleLike}
        />
      </>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  setErrorMsg: error => dispatch(setErrorMsg(error))
});

export default connect(undefined, mapDispatchToProps)(Player);
