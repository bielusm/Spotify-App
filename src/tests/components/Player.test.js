import React from "react";
import { shallow } from "enzyme";
import { Player } from "../../components/Player";
import * as PlayerApi from "../../api/player";

let bearerToken, is_playing, likes_song, id, update, wrapper;

beforeAll(() => {
  PlayerApi.prevTrack = jest.fn().mockResolvedValue();
  PlayerApi.pausePlayer = jest.fn().mockResolvedValue();
  PlayerApi.startPlayer = jest.fn().mockResolvedValue();
  PlayerApi.nextTrack = jest.fn().mockResolvedValue();
  PlayerApi.modifyLibrary = jest.fn().mockResolvedValue();
  update = jest.fn();
});

beforeEach(() => {
  jest.clearAllMocks();

  bearerToken = "12345";
  is_playing = false;
  likes_song = false;
  id = "12345";

  wrapper = shallow(
    <Player
      bearerToken={bearerToken}
      is_playing={is_playing}
      likes_song={likes_song}
      id={id}
      update={update}
    />
  );
});

test("Test Player component with default values", () => {
  expect(wrapper).toMatchSnapshot();
});

test("Test Player component with set values", () => {
  wrapper.setProps({ is_playing: true, likes_song: true });
  expect(wrapper).toMatchSnapshot();
});

test("Test prev button", async () => {
  await wrapper.find(".prevBtn").simulate("click");
  expect(PlayerApi.prevTrack).toHaveBeenCalledWith(bearerToken);
  expect(update).toHaveBeenCalled();
});

test("Test next button", async () => {
  await wrapper.find(".nextBtn").simulate("click");
  expect(PlayerApi.nextTrack).toHaveBeenCalledWith(bearerToken);
  expect(update).toHaveBeenCalled();
});

test("Test play/pause button", async () => {
  const btn = wrapper.find(".playBtn");

  wrapper.setProps({ is_playing: false });
  await btn.simulate("click");
  expect(PlayerApi.startPlayer).toHaveBeenCalledWith(bearerToken);
  expect(update).toHaveBeenCalled();

  wrapper.setProps({ is_playing: true });
  await btn.simulate("click");
  expect(PlayerApi.pausePlayer).toHaveBeenCalledWith(bearerToken);
  expect(update).toHaveBeenCalled();
});

test("Test like button", async () => {
  const btn = wrapper.find(".likeBtn");

  await btn.simulate("click");
  expect(PlayerApi.modifyLibrary).toHaveBeenCalledWith(
    bearerToken,
    id,
    likes_song
  );
  expect(update).toHaveBeenCalled();
});
