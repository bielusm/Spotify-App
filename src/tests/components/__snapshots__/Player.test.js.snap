// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Player component with default values 1`] = `
<Fragment>
  <WithStyles(ForwardRef(IconButton))
    className="prevBtn fas fa-backward"
    onClick={[Function]}
  />
  <WithStyles(ForwardRef(IconButton))
    className="playBtn fas fa-play"
    onClick={[Function]}
  />
  <WithStyles(ForwardRef(IconButton))
    className="nextBtn fas fa-forward"
    onClick={[Function]}
  />
  <WithStyles(ForwardRef(IconButton))
    className="likeBtn far fa-heart"
    onClick={[Function]}
  />
</Fragment>
`;

exports[`Test Player component with set values 1`] = `
<Fragment>
  <WithStyles(ForwardRef(IconButton))
    className="prevBtn fas fa-backward"
    onClick={[Function]}
  />
  <WithStyles(ForwardRef(IconButton))
    className="playBtn fas fa-pause"
    onClick={[Function]}
  />
  <WithStyles(ForwardRef(IconButton))
    className="nextBtn fas fa-forward"
    onClick={[Function]}
  />
  <WithStyles(ForwardRef(IconButton))
    className="likeBtn fas fa-heart"
    onClick={[Function]}
  />
</Fragment>
`;
